### PR TITLE
Igor lig 1788 smog add examples

### PR DESCRIPTION
**Pull request by IgorSusmelj**
## Changes

- add SMoG pytorch example
- add SMoG to readme

Changes by @philippmwirth:
- Refactor SMoG to use parameters instead of tensors
- Clean up benchmarks
- Add SMoG pytorch-lightning example

Note: I did not write a multi-gpu example because I don't think our implementation can handle it atm.

### CODE REVIEW

**Comment by IgorSusmelj** [comment]
Handed over PR to @philippmwirth

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 1]
```diff
@@ -841,16 +844,16 @@ def configure_optimizers(self):
...
-    BYOLModel,
-    DCL,
-    DCLW,
-    DINOModel,
-    MocoModel,
-    NNCLRModel,
-    SimCLRModel,
-    SimSiamModel,
-    SwaVModel,
+    # BarlowTwinsModel,
```
Comment (IgorSusmelj) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:847]: That's still a leftover of testing?

File: lightly/models/modules/heads.py [hunk 2]
```diff
@@ -232,34 +232,53 @@ def __init__(
...
         """
         assignments = self.assign_groups(x)
-        self.group_features = self.group_features.detach()
+        group_features = torch.clone(self.group_features.data)
         for assigned_class in torch.unique(assignments): 
             mask = assignments == assigned_class
-            self.group_features[assigned_class] = self.beta * self.group_features[assigned_class] + (1 - self.beta) * x[mask].mean(axis=0)
+            group_features[assigned_class] = self.beta * self.group_features[assigned_class] + (1 - self.beta) * x[mask].mean(axis=0)
+
+        assert group_features.requires_grad
```
Comment (IgorSusmelj) [inline lightly/models/modules/heads.py:275]: Sure we want to keep this in?

File: lightly/models/modules/heads.py [hunk 3]
```diff
@@ -232,34 +232,53 @@ def __init__(
...
+        return group_features
 
-        self.group_features= nn.functional.normalize(self.group_features, dim=1)
+    def set_group_features(self, x: torch.Tensor) -> None:
+        """Sets the group features and asserts they don't require gradient. """
+        self.group_features.data = x.to(self.group_features.device)
+        assert not self.group_features.requires_grad
```
Comment (IgorSusmelj) [inline lightly/models/modules/heads.py:281]: Same here

**Review by IgorSusmelj — COMMENTED** [review]
Haha, I neither can't approve nor request changes as I'm the author^^

I left a few comments. We're still missing the other two examples (pytorch lightning and distributed)

**Comment by philippmwirth** [comment]
> Haha, I neither can't approve nor request changes as I'm the author^^
> 
> I left a few comments. We're still missing the other two examples (pytorch lightning and distributed)

Hm somehow I didn't push the pytorch lightning example. I decided that it doesn't make sense to have a distributed example if `SMoG` is not implemented for distributed setups yet.

**Comment by philippmwirth** [comment]
> > I decided that it doesn't make sense to have a distributed example if SMoG is not implemented for distributed setups yet.
> 
> Ok, then I'd make a note to page (if users click on the smog distributed button) that this is not supported yet and will follow soon. And please don't forget to create the follow up issue so we don't forget about this.

There is no more distributed button.

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py (hunk 1 above)
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks/cifar10_benchmark.py:847]: Yes, I'll remove it before merging 👍

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 4]
```diff
@@ -841,16 +844,16 @@ def configure_optimizers(self):
...
+    # BYOLModel,
+    # DCL,
+    # DCLW,
+    # DINOModel,
+    # MocoModel,
+    # NNCLRModel,
+    # SimCLRModel,
+    # SimSiamModel,
+    # SwaVModel,
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:856]: Uncomment

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 5]
```diff
@@ -870,8 +869,8 @@ def __init__(self, dataloader_kNN, num_classes):
         utils.deactivate_requires_grad(self.projection_head_momentum)
 
         # smog
-        self.n_groups = 300 # 2% malus vs optimal setting of 3000 groups
-        memory_bank_size = 300 * batch_size # because we reset the group features every 300 iterations
+        self.n_groups = 300
+        memory_bank_size = 10000
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/imagenette_benchmark.py:873]: Why is this no longer 300 * batch_size ?

**Review by guarin — COMMENTED** [review]
Code and examples look great! I am running the examples and will approve once they ran through.

**Review by guarin — APPROVED** [review]
Examples ran through and loss decreases 💯

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py (hunk 5 above)
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:873]: Because `300 * batch_size >> n_samples`
